res/controllers/Numark-Scratch-scripts.js

```diff
@@ -0,0 +1,439 @@
+        }
+    };
+
+    // switch values are:
+    // 0 - switch in the middle
+    // 1 - switch up
+    // 2 - switch down
+    this.enableSwitch = function(channel, control, value, _status, _group) {
+        this.isSwitchHoldOn = value !== 0;
+        engine.setValue("[EffectRack1_EffectUnit1]", "group_[Channel" + deckNumber + "]_enable", (value !== 0));
+        engine.setValue("[EffectRack1_EffectUnit2]", "group_[Channel" + deckNumber + "]_enable", (value !== 0));
+        this.updateEffects();
+    };
+
+    this.dryWetKnob = new components.Pot({
+        group: "[EffectRack1_EffectUnit" + deckNumber + "]",
+        shift: function() {
+            this.inKey = "super1";
+        },
+        unshift: function() {
+            this.inKey = "mix";
+        },
+    });
+
+    this.effect1 = function(channel, control, value, status, _group) {
+        if (value === 0x7F) {
+            if (!NumarkScratch.shifted) {
+                NumarkScratch.allEffectOff();
+            }
+            this.effects[0] = !this.effects[0];
+            midi.sendShortMsg(status, control, this.effects[0] ? 0x7F : NumarkScratch.LOW_LIGHT);
+        }
+
+        this.updateEffects();
+    };
+
+    this.effect2 = function(channel, control, value, status, _group) {
+        if (value === 0x7F) {
+            if (!NumarkScratch.shifted) {
+                NumarkScratch.allEffectOff();
+            }
+            this.effects[1] = !this.effects[1];
+            midi.sendShortMsg(status, control, this.effects[1] ? 0x7F : NumarkScratch.LOW_LIGHT);
+        }
+
+        this.updateEffects();
+    };
+
+    this.effect3 = function(channel, control, value, status, _group) {
+        if (value === 0x7F) {
+            if (!NumarkScratch.shifted) {
+                NumarkScratch.allEffectOff();
+            }
+            this.effects[2] = !this.effects[2];
+            midi.sendShortMsg(status, control, this.effects[2] ? 0x7F : NumarkScratch.LOW_LIGHT);
+        }
+        this.updateEffects();
+    };
+};
+NumarkScratch.EffectUnit.prototype = new components.ComponentContainer();
+
+NumarkScratch.crossfader.setCurve = function(channel, control, value, _status, _group) {
+    switch (value) {
+    case 0x00: // Additive/Linear
+        engine.setValue("[Mixer Profile]", "xFaderMode", 0);
+        engine.setValue("[Mixer Profile]", "xFaderCalibration", 0.4);
+        engine.setValue("[Mixer Profile]", "xFaderCurve", 0.9);
+        break;
+    case 0x7F:  // Picnic Bench/Fast Cut
+        engine.setValue("[Mixer Profile]", "xFaderMode", 0);
+        engine.setValue("[Mixer Profile]", "xFaderCalibration", 0.9);
+        engine.setValue("[Mixer Profile]", "xFaderCurve", 7.0);
+    }
+};
+
+NumarkScratch.crossfader.xFaderReverse = function(channel, control, value, _status, _group) {
+    // 0x7F is ON, 0x00 is OFF
+    engine.setValue("[Mixer Profile]", "xFaderReverse", (value === 0x7F) ? 1 : 0);
+};
```

Comment:
this breaks on startup because the crossfader is only created after only in the init, so you'll need to move that into the init as well. The better alternative would probably be to crate your own Crossfader component outside the init and then instantiate that within the init.

Reply:
Figured this out, the sysex message to get controller states  was not capturing all these crossfader switch states, all other knobs report correctly

Reply:
I don't understand. The previous comment was about the fact that the CI fails because of a typeerror since the crossfader is created after this code is already executed.

Reply:
Ignore my previous comment, I think I might have a bug about something else. I'll debug after sorting this out.
Quick question, how is the section at [DJ505](https://github.com/mixxxdj/mixxx/blob/a5ae80ac68005dc34f9dcc8030d80035b8053238/res/controllers/Roland_DJ-505-scripts.js#L333), different to what I have, btw I got inspiration from that mapping. Trying to learn how that works on the 505 vs this.
Thanks for the help, as a non-coder it's much appreciated

Reply:
The same (potential) issue applies to the 505 as well, it's not in any way different... I guess we're lucky that it works there, but I wouldn't count on it, so that's why I'm suggesting to do it better here.

Reply:
Done